### DISCUSSION TITLE
Adventurer 5M Klipper Mod v00.04-beta

### DISCUSSION THREAD

**Discussion by xblax** [xblax/flashforge_ad5m_klipper_mod#101] [Announcements]
Unoffical mod for Flashforge Adventurer 5M (Pro) 3D printers to run Moonraker, custom Klipper, Mainsail & Fluidd. 
The Klipper Mod for ADM5 (Pro) is currently provided in two variants:

- "Lite" variant without on-screen GUI control application
- "KlipperScreen" variant. If you encounter any issues that could be caused by resource exhaustion (mostly system RAM), please check if the issue also occurs with the default variant.

Before installation read through the [Readme](https://github.com/xblax/flashforge_adm5_klipper_mod/blob/master/README.md) and other provided documentation.

### Changes
* klipper: fix runout sensor and implement filament runout handling (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/50)
* klipper: move calibrateable settings from pinter.base.cfg to [printer.cfg](https://github.com/xblax/flashforge_adm5_klipper_mod/blob/master/printer_configs/printer.base.cfg) (allow config save, issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/89)
* klipper: calibrated extruder rotational_distance - default 4.7 changed to 4.4 (discussion https://github.com/xblax/flashforge_adm5_klipper_mod/discussions/70)
* klipper: support pause/resume via `M24`/`M25` (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/82)
* klipper: enable stepper driver fan at 80% speed (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/83)
* klipper: improved `AUTO_BED_LEVEL` macro (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/73)
* klipper: support for `M600` filament change (https://github.com/xblax/flashforge_adm5_klipper_mod/discussions/91)
* klipper/klipperscreen: reworked `LOAD_FILAMENT` macro, interactive `LOAD_MATERIAL` macro (issues https://github.com/xblax/flashforge_adm5_klipper_mod/issues/16, https://github.com/xblax/flashforge_adm5_klipper_mod/issues/52)
* klipperscreen: feature is now integrated into master branch (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/32)
* klipperscreen: disable backlight in screensaver mode (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/47)
* klipperscreen: use tslib with touch calibration (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/46)
* klipperscreen: add timezone support (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/75)
* klipperscreen: fixes for font and theme settings (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/90)
* klipperscreen: improved wifi integration (issues https://github.com/xblax/flashforge_adm5_klipper_mod/issues/38, https://github.com/xblax/flashforge_adm5_klipper_mod/discussions/69)
* system: add nfs client tools (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/45)
* camera: improve compatibility with camera models (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/86)
* build system: added top level Makefile and reworked build scripts (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/32)
* pro: macros for air filtration settings, fix fan naming (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/68)
* pro: support for printer shutdown via Macro `SHUTDOWN` and physical button (issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues/56)

### Known issues
* 5M Pro: the internal fan door is installed incorrectly #54, #68 Note: this is a factory issue and not the mod!
* KlipperScreen: Wifi IP might not show up (#39)

---
> [!WARNING]
> *Make sure to backup your data and changed config files before [updating](https://github.com/xblax/flashforge_adm5_klipper_mod/blob/master/docs/INSTALL.md#updating)!*

> [!CAUTION]
> *If you want to install this mod to your ADM5 (Pro) then be aware, that you risk to loose your warranty or damage the printer. Proceed at your own risk if you want to try this mod!*

<hr /><em>This discussion was created from the release <a href='https://github.com/xblax/flashforge_adm5_klipper_mod/releases/tag/v00.04-beta'>Adventurer 5M Klipper Mod v00.04-beta</a>.</em>

**Comment by Tiwatz** [comment]
Great job!
I previously used the "lite" variant. Now I want to install the Klipper Screen variant, do I need to uninstall the old one or just install this over the old one?

**Reply by xblax** [reply]
Can just be updated or re-installed from one to the other. No need to uninstall first.

**Comment by DougJoseph** [comment]
Thanks! So cool! I'm currently printing my first job with the new v00.04-beta. Will report back how it all goes after a few print jobs, etc.

**Reply by gkoylu** [reply]
Hi, is there any problem while printing ? Did you face any issue with printer too ?

**Reply by DougJoseph** [reply]
It's been great except I've had only one issue. During one print, I "woke up" the LCD to check progress, and within about a minute or so after that, the Klipperscreen interface on the LCD reported that the MCU had run out of resources and the print was stalled and lost. All the many other prints completed flawlessly.

**Reply by gkoylu** [reply]
ooh, great.  

out ouf recources .... this is ram issue i think. Not enough space.  Also, i saw klipper fw on flashforge download center for 5m series. What do u think about that ?

**Reply by DougJoseph** [reply]
> Also, i saw klipper fw on flashforge download center for 5m series. What do u think about that ?

All the AD5M printers come from Flashforge with Klipper firmware on the mainboard. By license requirements they are obligated to published the Klipper firmware code they used on the mainboard. This mod does not touch that. It simply adds the missing parts they left off of the LCD: Moonraker, Mainsail, Fluidd, and optionally Klipperscreen. For remote access and control you need at least Moonraker + Mainsail, or Moonraker + Fluidd. For LCD access and control you need Klipperscreen too. That last part is what seems to be nudging things toward the resource limit on the LCD. But without it, no LCD access/control.

**Reply by consp** [reply]
> within about a minute or so after that, the Klipperscreen interface on the LCD reported that the MCU had run out of resources and the print was stalled 

Interesting and unfortunate to hear, not completely surprised though. Klipperscreen eats a bit too much resources and while I can probably strip some off it's a bit too much for the system. It's the main reason I started looking into guppyscreen which is less feature rich but uses a lot less and should not present those issues.

It's interesting it was caused by waking up since it doesn't actually sleep or anything. It might be the enabling of the backlight causing the issue (it's a separate script).

**Reply by DougJoseph** [reply]
> It's interesting it was caused by waking up since it doesn't actually sleep or anything.

It may have been merely a coincidence that it happened sometime after waking the screen. I should have grabbed a photo of the exact screen at the time, as it may have provided more info.

**Comment by gkoylu** [comment]
Hi,

I'm newbie on ff 5m pro. But i love klipper on my other custom sapphire plus since 3 years. So, That is the good news to get klipper on this amazing machine.   I was using TAILSCALE on my raspi based klipper  to reach my dashboard over Wide area network with a static wan ip address. So, can we handle about this in somehow ? Just an opinion.

With my best regards,

**Reply by consp** [reply]
Since it's a simple VPN, it should be possible, you have to create a route  from the PI to the machine or connect to the IP of the printer (depends on how your vpn is set up).

**Reply by gkoylu** [reply]
my printer connected a modem as usually. What can i do for routing ?

**Reply by consp** [reply]
What do you mean by that, is it connected to the modem/router/switch or directly to the internet (NOT RECOMMENDED, it's not secure).

**Reply by gkoylu** [reply]
Yes, connected to modem

**Comment by RikshaDriver** [comment]
FYI printer configs from 5M Pro Factory image ( v2.4.5 - https://file.io/v4gHasvLUbtl ) appear different to printer .config files on this repo.

Are these enhanced or reverse engineered?

**Reply by consp** [reply]
> Are these enhanced or reverse engineered?

Settings are taken from the originals, when they make sense. Some things have been moved to printer.cfg to make them more adjustable with the software used.

Mostly improvements, renaming to make it more logical and the macros are moved:

```
diff printer.base.cfg.mod printer.base.cfg
19,20c19,20
< position_max: 110.1
< position_min: -110
---
> position_max: 125
> position_min: -125.0
22c22
< homing_retract_dist: 5
---
> homing_retract_dist: 10
24c24
< second_homing_speed: 5
---
> #homing_positive_dir: false
34,35c34,35
< position_max: 110.1
< position_min: -110
---
> position_max: 125.0
> position_min: -125.0
37c37
< homing_retract_dist: 5
---
> homing_retract_dist: 10
39c39
< second_homing_speed: 5
---
> #homing_positive_dir: false
52,55c52,55
< homing_speed:25
< homing_retract_dist: 5
< homing_retract_speed: 12.5
< second_homing_speed: 5
---
> homing_speed:20
> #homing_positive_dir: false 
> homing_retract_dist: 5  
> homing_retract_speed: 10
61a62
> rotation_distance: 4.7
66a68,71
> control: pid
> pid_Kp: 20.109
> pid_Ki: 1.943
> pid_Kd: 52.031
78a84,87
> control: pid
> pid_Kp: 32.79
> pid_Ki: 4.970
> pid_Kd: 54.118
95a105
> 
98,99c108
< hardware_pwm: False
< initial_white: 0.8
---
> hardware_pwm: False 
101c110
< [heater_fan hotend_fan]
---
> [heater_fan heat_fan]
108c117
< [fan]
---
> [fan_generic fanM106]                                           
116c125
< [controller_fan stepper_driver_fan]
---
> [fan_generic pcb_fan]
118,120d126
< stepper: stepper_x, stepper_y, stepper_z
< fan_speed: 0.8
< idle_timeout: 0
123c129
< pin:PB6
---
> pin:PB8
126c132,133
< pin:PB8
---
> pin:PB6
> 
128,129d134
< # unclear what the "power" pins do at the moment
< # disable them in the UI for now
133c138
<     _SHUTDOWN_BUTTON_TRIGGER
---
>     M105
135c140
< [output_pin _power_off]
---
> [output_pin power_off]
140c145
< [output_pin _clear_power_off]
---
> [output_pin clear_power_off]
210c215,229
< [temperature_sensor Load_Cell]
---
> 
> [output_pin level_h1]
> pin:PG2
> value:0
> 
> [output_pin level_h2]
> pin:PG3
> value:0
> 
> [output_pin level_h3]
> pin:PG4
> value:0
> 
> 
> [temperature_sensor weightValue]
222c241
< [temperature_sensor TVOC_Level]
---
> [temperature_sensor tvocValue]
235,238d253
< # reset the check_level_pin state
< [output_pin _level_clear]
< pin: PD10
< value:0
240,254c255,257
< # check level pin is triggered when load cell tare is set via level_h1 pin
< [gcode_button _check_level_pin]
< pin: !PE0
< press_gcode:
<     G4 P0
< 
< # set load cell tare
< [output_pin _level_h1]
< pin:PG2
< value:0
< 
< # don't use (!) - changes load cell calibration
< [output_pin _level_h2]
< pin:PG3
< value:0
---
> [output_pin level_clear]                                               
> pin: PD10
> value:0    
256,259c259,263
< # don't use (!) - changes load cell calibration
< [output_pin _level_h3]
< pin:PG4
< value:0
---
> #Right extruder filament sensor
> [filament_switch_sensor e1_sensor]
> pause_on_runout: False
> switch_pin: PE1
> event_delay: 1.0
261,266c265,268
< # Unused Filament runout sensor e1 must be configure as an input pint 
< # Otherwise the runout sensor does not work
< [gcode_button _e1_sensor_input]
< pin: !PE1
< press_gcode:
<     G4 P0
---
> [gcode_button check_level_pin]            
> pin: !PE0                                 
> press_gcode:                     
>     M105 
268,269c270,271
< # Filament runout sensor
< [filament_switch_sensor runout_sensor]
---
> #Right extruder filament sensor
> [filament_switch_sensor e0_sensor]
273,274d274
< runout_gcode:
<     _FILAMENT_RUNOUT_EVENT
281c281
< initial_angle:95
---
> initial_angle:90
286,289d285
<   
< [virtual_sdcard]
< path: /root/printer_data/gcodes
< on_error_gcode: CANCEL_PRINT
291c287,295
< [display_status]
---
> [gcode_macro M900]
> gcode:
>   {% if 'K' in params %}
>     {% if 'E' in params %}
>       SET_PRESSURE_ADVANCE EXTRUDER={params.E} ADVANCE={params.K}
>     {% else %}
>       SET_PRESSURE_ADVANCE ADVANCE={params.K}
>     {% endif %}
>   {% endif %}
293c297,318
< [exclude_object]
---
> [gcode_macro M106]
> gcode:
>   {% if 'S' in params %}
>     {% set speed = params['S']|float %}
>   {% else %}
>     {% set speed = 255.0|float%}
>   {% endif %}
>   {% if 'T' in params %}
>     {% set index = params['T']|int %}
>   {% endif %}
>   {% if 'P' in params %}
>     {% set index = params['P']|int %}
>   {% endif %}
>   {% if index == 101 %}
>       SET_FAN_SPEED FAN=chamber_fan SPEED={speed/255.0}
>   {% elif index == 2 %}
>       SET_FAN_SPEED FAN=chamber_fan SPEED={speed/255.0}
>   {% elif index == 102 %}
>       
>   {% else %}
>       SET_FAN_SPEED FAN=fanM106 SPEED={speed/255.0}
>   {% endif %}
295,299c320
< # Home Axis in order Z, X, Y
< # Move away from endstop to prevent damage to plastic tab in bed corner
< # if bed is raised to nozzle and don't keep head or bed pressed
< # against endstop for long time
< [homing_override]
---
> [gcode_macro M107]
301,307c322,339
<     # Limit acceleration for homing and moves on print start
<     M204 S5000
<     G28 Z0
<     G28 X0
<     G28 Y0
<     G90
<     G1 X105 Y105 Z210 F1800
---
>   {% if 'T' in params %}
>     {% set index = params['T']|int %}
>   {% endif %}
>   {% if 'P' in params %}
>     {% set index = params['P']|int %}
>   {% endif %}
>   {% if index == 101 %}
>       SET_FAN_SPEED FAN=chamber_fan SPEED=0.0
>   {% elif index == 2 %}
>       SET_FAN_SPEED FAN=chamber_fan SPEED=0.0
>   {% else %}
>       SET_FAN_SPEED FAN=fanM106 SPEED=0.0
>   {% endif %}
>   
> [virtual_sdcard]
> path: /data/
> 
> [display_status]
312,313c344,345
< horizontal_move_z:2
< move_check_distance: 5
---
> horizontal_move_z:5              
> move_check_distance: 8             
324,327c356,359
< [probe]
< pin: !PB4
< speed: 2
< lift_speed: 5
---
> [probe]                            
> pin: !PB4 
> z_offset: -0.25       
> speed: 5                   
328a361,362
> 
> 
```

**Reply by gkoylu** [reply]
is it possible to turn off light to end-gcode section at the slicer ?

**Reply by Tiwatz** [reply]
> is it possible to turn off light to end-gcode section at the slicer ?

You can add LED control macros:
https://github.com/xblax/flashforge_ad5m_klipper_mod/discussions/130
Then use LED_OFF in your slicer end gcode

**Comment by staniox** [comment]
when i put a file named klipper_mod_skip on a USB drive and plug it in before printer start it doesn't work any solution?
